### PR TITLE
Fix image match

### PR DESCRIPTION
**Pull request by jimmyxian**
Origin Issue:
https://github.com/docker/swarm/issues/740

Replace:
https://github.com/docker/swarm/pull/791

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @vieux 
This PR is what i'm thinking about.
https://github.com/docker/swarm/pull/791#issuecomment-103011312
```
The following two interfaces have different behavior.
docker rmi: If remove one image with no tag, it will remove latest.
docker save: If save one image with no tag, it will save all the images with repo name.

This two interface in swarm both use image.Match(). I think it's not proper.
In rmi, match image with default latest tag if rmi with no tag.
In docker save, If save with no tag, match image without tag, just match repo name.

So, two functions maybe OK. Match and MatchWithoutTag()
```

File: api/router.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ var routes = map[string]map[string]handler{
 		"/images/viz":                     notImplementedHandler,
 		"/images/search":                  proxyRandom,
 		"/images/get":                     getImages,
-		"/images/{name:.*}/get":           proxyImage,
+		"/images/{name:.*}/get":           getImage,
```
Comment (vieux) [inline api/router.go]: @jimmyxian I don't get why do you need a custom handler here, using proxy ensure we have the same behaviour as the docker daemon.

Comment (jimmyxian) [inline api/router.go]: @vieux Actually, not a custom handler. The difference is the `image match`
```
+// GET /images/{name:.*}/get
+func getImage(c *context, w http.ResponseWriter, r *http.Request) {
+ name := mux.Vars(r)["name"]
+
+     for _, image := range c.cluster.Images() {
+         if len(strings.SplitN(name, ":", 2)) == 2 && image.Match(name) ||
+           len(strings.SplitN(name, ":", 2)) == 1 && image.MatchWithoutTag(name) {
+                proxy(c.tlsConfig, image.Engine.Addr, w, r)
+                return
+          }
+     }
+     httpError(w, fmt.Sprintf("No such image: %s", name), http.StatusNotFound)
+}
```

**Comment by jimmyxian** [comment]
ping @vieux 
Done to merge two functions. Check again, thanks

**Comment by vieux** [comment]
LGTM

**Comment by vieux** [comment]
ping @abronan

**Comment by aluzzardi** [comment]
@vieux The code looks good to me but I'm not familiar at all with the logic so if you say it's good let's merge it

**Pull request merged**